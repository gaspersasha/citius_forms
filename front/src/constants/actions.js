export const USER_LOGIN = 'USER_LOGIN';
export const USER_LOGOUT = 'USER_LOGOUT';
export const USER_SET_STATE = 'USER_SET_STATE';

export const DETAILS_GET = 'DETAILS_GET';
export const DETAILS_POST = 'DETAILS_POST';
export const DETAILS_SET_STATE = 'DETAILS_SET_STATE';

export const TOGGLE_ORDER_SUMMARY = 'TOGGLE_ORDER_SUMMARY';

export const CONTACT_DEALER_POST = 'CONTACT_DEALER_POST';
export const CONTACT_DEALER_SET_STATE = 'CONTACT_DEALER_SET_STATE';
export const CONTACT_DEALER_GET_VEHICLE = 'CONTACT_DEALER_GET_VEHICLE';
