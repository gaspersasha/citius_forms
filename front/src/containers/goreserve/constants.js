export const PAYMENT_METHOD = {
  stripe: 'open-stripe-form',
  payPal: 'open-paypal',
};

export const PAYMENT_METHODS_OPTIONS = [
  {
    value: 'cash',
    option: 'I will buy outright',
  },
  {
    value: 'finance',
    option: 'I will buy on finance',
  },
];
