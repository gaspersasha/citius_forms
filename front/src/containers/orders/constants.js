// The note: type can be null
export const PAYMENT_TYPE = {
  DEBIT: 'debit',
  CREDIT: 'credit',
  PAID: 'paid',
};

export const ORDER_SECTION_TYPE = {
  financialSummary: {
    title: 'Financial summary',
    subtitle: 'Deposit payments',
  },
  yourOrders: {
    title: 'Your orders',
    subtitle: 'Orders',
  },
};
