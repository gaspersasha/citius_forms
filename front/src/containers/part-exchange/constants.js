export const CAR_CONDITION_OPTIONS = [
  { value: 'great', option: 'Great' },
  { value: 'good', option: 'Good' },
  { value: 'fair', option: 'Fair' },
  { value: 'poor', option: 'Poor' },
];

export const SERVICE_HISTORY_OPTIONS = [
  { value: 'full', option: 'Full' },
  { value: 'part', option: 'Part' },
  { value: 'notdue', option: 'Not due' },
  { value: 'none', option: 'None' },
];
