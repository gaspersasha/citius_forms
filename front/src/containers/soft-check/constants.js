export const KEYS = [
  'title',
  'firstName',
  'middleName',
  'lastName',
  'email',
  'phone',
  'dateOfBirth',
  'postCode',
  'houseName',
  'houseNumber',
  'street',
  'town',
  'county',
  'currentIncome',
  'district',
];
export const MAX_SALARY = 9999999999.99;
export const SEARCH_PRICE_TAGS_LIST = [100, 150, 200, 300, 500];
