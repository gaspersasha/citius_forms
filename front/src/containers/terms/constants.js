export const MILAGE_OPTIONS = [
  { value: 6000, option: '6,000 miles' },
  { value: 7000, option: '7,000 miles' },
  { value: 8000, option: '8,000 miles' },
  { value: 9000, option: '9,000 miles' },
  { value: 10000, option: '10,000 miles' },
  { value: 11000, option: '11,000 miles' },
  { value: 12000, option: '12,000 miles' },
  { value: 13000, option: '13,000 miles' },
  { value: 14000, option: '14,000 miles' },
  { value: 15000, option: '15,000 miles' },
  { value: 16000, option: '16,000 miles' },
  { value: 17000, option: '17,000 miles' },
  { value: 18000, option: '18,000 miles' },
  { value: 19000, option: '19,000 miles' },
  { value: 20000, option: '20,000 miles' },
  { value: 21000, option: '21,000 miles' },
  { value: 22000, option: '22,000 miles' },
  { value: 23000, option: '23,000 miles' },
  { value: 24000, option: '24,000 miles' },
  { value: 25000, option: '25,000 miles' },
  { value: 26000, option: '26,000 miles' },
  { value: 27000, option: '27,000 miles' },
  { value: 28000, option: '28,000 miles' },
  { value: 29000, option: '29,000 miles' },
  { value: 30000, option: '30,000 miles' },
];

export const MONTHS_OPTIONS = [
  { value: 24, option: '24 months' },
  { value: 36, option: '36 months' },
  { value: 48, option: '48 months' },
  { value: 60, option: '60 months' },
];

export const FINANCE_OPTIONS = [
  { value: 'PCP', option: 'PCP' },
  { value: 'HP', option: 'HP' },
];

export const DELIVERY_INFO = [
  {
    value: 18,
    option: 'Driven delivery within England and Wales 239.0',
  },
  {
    value: 19,
    option: 'Driven/Ferry delivery to Belfast 599.0',
  },
  {
    value: 20,
    option: 'Driven delivery to Scotland 355.0',
  },
  {
    value: 21,
    option: 'Transporter delivery within England and Wales 479.0',
  },
  {
    value: 22,
    option: 'Transporter delivery to Scotland 599.0 ',
  },
];

export const FINANCE_CREDIT_RATING_OPTIONS = [
  { value: 'excellent', option: 'Excellent' },
  { value: 'good', option: 'Good' },
  { value: 'average', option: 'Average' },
  { value: 'poor', option: 'Poor' },
];

export const FINANCE_TYPE_CONTENT = {
  PCP: {
    title: '(PCP) Personal Contract Purchase',
    subtitle:
      'You make low monthly payments. At the end you either return the cars or buy it for an additional fee, often known as a baloon payment.',
    description:
      ' Personal Contract Purchase is a finance plan that allows you to defer part of the cost of the vehicle until the end of the agreement, reducing the regular monthly payments. The deferred element known as the guaranteed future value or optional final payment is based upon the likely value of the vehicle at the end of the agreement and is set by the finance company based upon the annual mileage selected. At the end of the term, you have the option to hand the vehicle back, part-exchange it or pay to final payment to take ownership of the vehicle. If you choose to hand the vehicle back excess mileage and damage charges may apply.',
    max: 40,
  },
  HP: {
    title: '(HP) Hire Purchase',
    subtitle:
      'You pay for the car in a series of installments. At the end of the agreed repayment period, it is yours.',
    description:
      'Hire Purchase is a finance plan that allows you to pay for the vehicle by fixed monthly payments over the term of the agreement. Hire purchase is a secured loan so the finance company has the vehicle as security for the loan until all payments and fees are made, at the end of the term of the agreement after all payments and option to purchase fee are made you take full ownership of the vehicle.',
    max: 80,
  },
};

export const NEW_FINANCE_QUOTE = {
  PCP: {
    capId: '49573',
    financeType: 'PCP',
    numRegularPayments: 46,
    paymentAmount: 152.16,
    apr: 8.9,
    finalPayment: 3173.34,
    gfv: 0,
    basicPrice: 9238.8,
    rateOfInterest: 4.61,
    flatRate: 0,
    totalCashPrice: 9238.8,
    firstPayment: 152.16,
    documentationFee: 0,
    purchaseFee: 0,
    totalPayable: 11248.86,
    totalCharges: 0,
    commission: 0,
    quoteReference:
      'https://services.codeweavers.net/forms/quote/5ba82157-a463-4754-b560-7385b88dd237',
    totalDeposit: 924,
    mileagePA: 8000,
    periodMonths: 36,
    amountFinanced: 8314.8,
    financeQuoteRequestDetails: 15,
    financeKey: 'ALPHERA_PCP',
  },
  HP: {
    capId: '49573',
    financeType: 'Hire Purchase',
    numRegularPayments: 58,
    paymentAmount: 170.8,
    apr: 8.9,
    finalPayment: 170.8,
    gfv: 0,
    basicPrice: 9238.8,
    rateOfInterest: 4.65,
    flatRate: 0,
    totalCashPrice: 9238.8,
    firstPayment: 170.8,
    documentationFee: 0,
    purchaseFee: 0,
    totalPayable: 11172,
    totalCharges: 0,
    commission: 0,
    quoteReference:
      'https://services.codeweavers.net/forms/quote/a7ccf346-ef3c-471f-b718-4a9c40606a99',
    totalDeposit: 924,
    mileagePA: 8000,
    periodMonths: 36,
    amountFinanced: 8314.8,
    financeQuoteRequestDetails: 26,
    financeKey: 'ALPHERA_HP',
  },
};
