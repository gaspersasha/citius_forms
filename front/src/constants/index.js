import * as ACTIONS from './actions';
import * as FORM from './form';
import * as URL from './urls';
import * as SCC from './scc';
import * as CONTACTS from './contacts';
import * as ENV from './env';

// TODO @ORDER_SUMMARY_DUMMY. Mock props
import * as ORDER_SUMMARY_DUMMY from './OrderSummaryDummy.json';

export { ACTIONS };
export { FORM };
export { URL };
export { CONTACTS };
export { ENV };
export { SCC };
export { ORDER_SUMMARY_DUMMY };
export { default as DEVICE_TYPES } from './device-types';
export { default as INPUT_STATUS } from './input-status';
export { default as VALIDATION_TYPES } from './validation-types';
