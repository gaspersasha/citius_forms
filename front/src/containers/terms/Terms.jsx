import React, { useState, useEffect } from 'react';
import { useRouter } from 'next/router';
import { useUserContext } from '~contexts';
import { INPUT_STATUS, VALIDATION_TYPES } from '~constants';
import {
  Button,
  Modal,
  Spinner,
  PrivacyNotice,
  Parts,
  ProgressBar,
  Seo,
} from '~components';
import {
  endpointFetch,
  endpointPush,
  validate,
  isValid,
  prepare,
  scrollTo,
  getFormData,
  makeQueryString,
  fillProgress,
  GTM_EVENT,
} from '~utils';
import TermsInputs from './TermsInputs';
import TermsResults from './TermsResults';
import { FINANCE_TYPE_CONTENT } from './constants';

import s from './terms.module.sass';

const Terms = () => {
  const [formsProgress, setFormsProgress] = useState([[1, 0]]);
  const [quotes, setQuotes] = useState({});
  const [maxFinanceDeposit, setMaxFinanceDeposit] = useState({
    value: 4400,
    status: INPUT_STATUS.DEFAULT,
    validationType: VALIDATION_TYPES.NUMBER,
  });
  const [state, setState] = useState({
    deposit: {
      value: '',
      status: INPUT_STATUS.DEFAULT,
      validationType: VALIDATION_TYPES.NUMBER,
    },
    mileage: {
      value: '',
      status: INPUT_STATUS.DEFAULT,
      validationType: VALIDATION_TYPES.NUMBER,
    },
    terms: {
      value: '',
      status: INPUT_STATUS.DEFAULT,
      validationType: VALIDATION_TYPES.NUMBER,
    },
    delivery: {
      value: 18,
      status: INPUT_STATUS.DEFAULT,
      validationType: VALIDATION_TYPES.NUMBER,
    },
    credit: {
      value: '',
      status: INPUT_STATUS.DEFAULT,
      validationType: VALIDATION_TYPES.NOT_EMPTY,
    },
  });
  const [financeType, setFinanceType] = useState({
    finance: {
      value: 'PCP',
      status: INPUT_STATUS.DEFAULT,
      validationType: VALIDATION_TYPES.NOT_EMPTY,
    },
  });
  const [carInfo, setCarInfo] = useState({ price: '' }); // price
  const [userInfo, setUserInfo] = useState({ creditScore: '' }); // creditScore | creditSearch
  const [isLoading, setIsLoading] = useState(false);
  const [isModalOpen, setIsModalOpen] = useState(false);
  const [apiError, setApiError] = useState('');
  const [modalType, setModalType] = useState('');
  const [deliveryOptions, setDeliveryOptions] = useState([]);
  const [unsubscribe, setUnsubscribe] = useState(false);

  const {
    query: { productAdvertId, quoteId, leadgen },
  } = useRouter();
  const router = useRouter();
  const isLeadGen = !!Number(leadgen);

  const { user } = useUserContext();

  useEffect(() => {
    const isUnsubscribe =
      user.subscribed !== undefined ? !user.subscribed : false;

    setUnsubscribe(isUnsubscribe);
  }, []);

  const onUnsub = () => setUnsubscribe(!unsubscribe);

  const financeQuoteRequest = (params) => {
    const formData = getFormData(params);

    return endpointPush('POST', 'newFinanceQuote', {}, formData);
  };

  const calcMaxFinanceDeposit = (type) =>
    Math.round(
      (FINANCE_TYPE_CONTENT[type].max / 100) * Number(carInfo.price || 100000)
    );

  const initialQuoteSuccess = (financeQuote, creditScore) => {
    let creditRating = '';

    if (!financeQuote) throw new Error('empty_quotes');

    if (creditScore >= 1 && creditScore <= 319) creditRating = 'poor';

    if (creditScore >= 320 && creditScore <= 409) creditRating = 'average';

    if (creditScore >= 410 && creditScore <= 469) creditRating = 'good';

    if (creditScore >= 470) creditRating = 'excellent';

    const { deposit, mileage, terms, delivery, credit } = state;

    if (Object.values(financeQuote)[0].financeType === 'Hire Purchase') {
      Object.values(financeQuote)[0].financeType = 'HP';
    }

    setState({
      ...state,
      deposit: {
        ...deposit,
        value: Object.values(financeQuote)[0].totalDeposit,
      },
      mileage: { ...mileage, value: Object.values(financeQuote)[0].mileagePA },
      terms: { ...terms, value: Object.values(financeQuote)[0].periodMonths },
      delivery: {
        ...delivery,
        value: Object.values(financeQuote)[0].deliveryMethodId || 18,
      },
      credit: { ...credit, value: creditRating || userInfo.creditScore },
    });
    setFinanceType({
      finance: {
        ...financeType.finance,
        value: Object.values(financeQuote)[0].financeType,
      },
    });
    setQuotes({ ...quotes, ...financeQuote });
    setMaxFinanceDeposit({
      ...maxFinanceDeposit,
      value: calcMaxFinanceDeposit(
        Object.values(financeQuote)[0].financeType || 'PCP'
      ),
    });
  };

  const setDeliveryInfo = (deliveries) => {
    deliveries.map((delivery) =>
      setDeliveryOptions((deliveryOption) => [
        ...deliveryOption,
        {
          value: delivery.deliveryMethodId,
          option: `${delivery.description} Â£${delivery.deliveryCostPlusVAT}`,
        },
      ])
    );
  };

  const fetchData = async () => {
    try {
      setIsLoading(true);
      const quoteInfo = await financeQuoteRequest({
        pageload: true,
        productAdvertId,
      });
      const userProfileInfo = await endpointFetch('userProfileInfo');
      const productInformation = await endpointFetch('productInformation', {
        productAdvertId,
      });
      const getDeliveryOptions = !isLeadGen
        ? await endpointFetch('getDeliveryOptions', {
            productAdvertId,
          })
        : null;

      setUserInfo({ ...userProfileInfo });

      initialQuoteSuccess(quoteInfo, userProfileInfo.creditScore);

      setCarInfo({ ...productInformation });

      if (!isLeadGen) {
        setDeliveryInfo(getDeliveryOptions);
      }
    } catch (error) {
      setApiError('');
    }

    setIsLoading(false);
  };

  useEffect(() => {
    fetchData();
  }, [productAdvertId]);

  const financeQuoteSuccess = (financeQuote) => {
    if (!financeQuote) throw new Error('empty_quotes');

    if (Object.values(financeQuote)[0].financeType === 'Hire Purchase') {
      Object.values(financeQuote)[0].financeType = 'HP';
    }

    setQuotes({ ...financeQuote });
    setApiError('');
    setMaxFinanceDeposit({
      ...maxFinanceDeposit,
      value: calcMaxFinanceDeposit(
        Object.values(financeQuote)[0].financeType || 'PCP'
      ),
    });

    if (Object.keys(quotes).length !== Object.keys(financeQuote).length) {
      setFinanceType({
        finance: {
          ...financeType.finance,
          value: Object.values(financeQuote)[0].financeType,
        },
      });
    }
  };

  const setErrorMessage = (error) => setApiError(error);

  useEffect(() => {
    const { price: totalPrice } = carInfo;
    const { deposit, terms, mileage, delivery, credit } = state;
    const { finance } = financeType;
    const params = {
      pageload: false,
      financeType: finance.value === 'HP' ? 'Hire Purchase' : finance.value,
      financeCreditRating: credit.value,
      totalPrice,
      deliveryMethodId: delivery.value,
      productAdvertId,
      financeDeposit: deposit.value,
      financePeriodMonths: terms.value,
      financeAnnualMileage: mileage.value,
    };

    if (deposit.status !== INPUT_STATUS.INVALID) {
      financeQuoteRequest(params)
        .then((data) => financeQuoteSuccess(data))
        .catch(setErrorMessage);
    }
  }, [
    state.deposit.value,
    state.mileage.value,
    state.terms.value,
    state.delivery.value,
    state.credit.value,
  ]);

  const handleChange = ({ target: { name, value, id } }) => {
    let status = INPUT_STATUS.DEFAULT;
    let type = value || id;
    let maxDeposit = '';

    switch (name || id) {
      case 'finance': {
        if (value === 'Hire Purchase') {
          type = 'HP';
        }

        maxDeposit = calcMaxFinanceDeposit(type);

        if (state.deposit.value > maxDeposit) {
          status = INPUT_STATUS.INVALID;
        }

        setState({
          ...state,
          deposit: { ...state.deposit, status },
        });
        setFinanceType({
          finance: {
            ...financeType.finance,
            value: type,
          },
        });
        setMaxFinanceDeposit({ ...maxFinanceDeposit, value: maxDeposit });

        break;
      }

      case 'deposit': {
        if (!value || value > maxFinanceDeposit.value) {
          status = INPUT_STATUS.INVALID;
        }

        maxDeposit = calcMaxFinanceDeposit(financeType.finance.value || 'PCP');

        setState({
          ...state,
          deposit: { ...state.deposit, value, status },
        });
        setMaxFinanceDeposit({ ...maxFinanceDeposit, value: maxDeposit });

        break;
      }

      case String(id): {
        if (id === 'Hire Purchase') {
          type = 'HP';
        }

        maxDeposit = calcMaxFinanceDeposit(type);

        if (state.deposit.value > maxDeposit) {
          status = INPUT_STATUS.INVALID;
        }

        setState({
          ...state,
          deposit: { ...state.deposit, status },
        });
        setFinanceType({
          finance: {
            ...financeType.finance,
            value: type,
          },
        });
        setMaxFinanceDeposit({ ...maxFinanceDeposit, value: maxDeposit });

        break;
      }

      default:
        setState({
          ...state,
          [name]: {
            ...state[name],
            value,
            status: INPUT_STATUS.DEFAULT,
          },
        });
    }
  };

  const setNextStep = (index, max, nextStep) => {
    setFormsProgress((prevState) => {
      const updateStatuses = [...prevState];

      updateStatuses[index] = [max, nextStep];

      return [...updateStatuses];
    });
  };

  const StepState = (index) => {
    const [max, curr] = formsProgress[index];
    const nextStep = curr + 1;

    setNextStep(index, max, nextStep);

    return {
      nextStep,
      isFilled: nextStep === max,
    };
  };

  const handleSubmit = (e) => {
    e.preventDefault();
    const validatedState = validate(prepare(state));

    setState(validatedState);
    const tryAgain = !isValid(validatedState);

    if (tryAgain) return scrollTo('.error-input-border');

    StepState(0);

    const newState = {
      ...quotes[String(financeType.finance.value)],
      quoteId,
      deliveryMethodId: state.delivery.value,
    };

    endpointPush('POST', 'updateFinanceTerms', newState).then(() => {
      GTM_EVENT('ssFinanceTermsCompleted');

      const url = makeQueryString('almost-done', {
        quoteId: router.query.quoteId || '',
        quoteItemId: router.query.quoteItemId || '',
        productAdvertId: router.query.productAdvertId || '',
      });

      router.push(url);
    });
  };

  const toggleModal = () => setIsModalOpen(!isModalOpen);

  const showModal = (type) => {
    setModalType(type);
    toggleModal();
  };

  const infoModal = (
    <Modal isDark onClose={toggleModal}>
      <div>
        {modalType === 'PCP' ? (
          <div>
            <span>{FINANCE_TYPE_CONTENT.PCP.title}</span>
            <p>{FINANCE_TYPE_CONTENT.PCP.subtitle}</p>
            <p>{FINANCE_TYPE_CONTENT.PCP.description}</p>
          </div>
        ) : (
          <div>
            <span>{FINANCE_TYPE_CONTENT.HP.title}</span>
            <p>{FINANCE_TYPE_CONTENT.HP.subtitle}</p>
            <p>{FINANCE_TYPE_CONTENT.HP.description}</p>
          </div>
        )}
      </div>
    </Modal>
  );

  const disabledForm = (type) => {
    if (
      state.terms.value > FINANCE_TYPE_CONTENT.PCP.max &&
      type === ('HP' || 'Hire Purchase')
    ) {
      return false;
    }
  };

  const disableFinanceDeposit =
    (maxFinanceDeposit.value < state.deposit.value ||
      state.deposit.value < 0) &&
    true;

  const disabledFinanceValue =
    !quotes.PCP && financeType.finance.value === 'PCP' && true;

  const disabledTrue = true;

  const progress = fillProgress(formsProgress);

  return (
    <div className={s.container}>
      <Seo />
      {isModalOpen && infoModal}
      <div className={s.controls}>
        <div className="align-corner">
          <button className={s.back} type="button" onClick={router.back}>
            Back
          </button>
          <span className={s.progress}>{`${progress}% Completed`}</span>
        </div>

        <ProgressBar progress={progress} />
        <Parts part={2} step={4} />
      </div>
      <div className="form-container">
        <div className="form">
          {isLoading || !productAdvertId ? (
            <Spinner />
          ) : (
            <>
              <h1 className={s.title}>Choose your finance terms</h1>
              <p className={s.subTitle}>
                Tailor your finance details to see how much you could be paying
              </p>
              <form className={s.financeForm} onSubmit={handleSubmit}>
                <TermsInputs
                  disabledTrue={disabledTrue}
                  fields={{ ...state, ...financeType }}
                  handleChange={handleChange}
                  deliveryOptions={deliveryOptions}
                  isLeadGen={isLeadGen}
                  maxFinanceDeposit={maxFinanceDeposit.value}
                />
                <div className={s.financeType}>
                  <h5>Now select your finance type</h5>
                  <TermsResults
                    disabledForm={disabledForm}
                    fields={{ ...state, ...financeType }}
                    quotes={quotes}
                    handleChange={handleChange}
                    showModal={showModal}
                    apiError={apiError}
                  />
                </div>
                <Button
                  type="submit"
                  className={s.submitBtn}
                  disabled={
                    disableFinanceDeposit || disabledFinanceValue || !!apiError
                  }
                >
                  Next
                </Button>
              </form>
              <div className={s.descriptions}>
                <p>Representative example when buying on PCP</p>
                <p>
                  Borrowing Â£9,500 over 48 months, zero deposit, on PCP, an
                  annual mileage of 8,000pa, with a Representative APR of 9.9%,
                  the amount payable would be Â£192 per month, an optional final
                  payment of Â£2,923, with a total cost of credit of Â£2,426 and a
                  total amount payable of Â£11,926 (** based on a Ford Fiesta
                  Titanium hatchback)
                </p>
                <p>
                  Buyacar is a credit broker, not a lender. Our rates start from
                  6.9% APR. The rate you are offered will depend on your
                  individual circumstances.
                </p>
              </div>
            </>
          )}
        </div>
        <PrivacyNotice onChange={onUnsub} unsubscribe={unsubscribe} />
      </div>
    </div>
  );
};

export default Terms;
