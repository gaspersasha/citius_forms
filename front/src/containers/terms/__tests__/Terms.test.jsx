import React from 'react';
import { useRouter } from 'next/router';
import { useUserContext, useGOContext, useVehicleContext } from '~contexts';
import { INPUT_STATUS } from '~constants';
import { endpointFetch, endpointPush } from '~utils';
import Terms from '../Terms';
import { DELIVERY_INFO, NEW_FINANCE_QUOTE } from '../constants';

jest.mock('~contexts', () => ({
  useUserContext: jest.fn(),
  useGOContext: jest.fn(),
  useVehicleContext: jest.fn(),
}));

jest.mock('next/router', () => ({
  useRouter: jest.fn(),
}));

useRouter.mockImplementation(() => ({
  query: { productAdvertId: '3362300' },
}));

jest.mock('~utils', () => ({
  ...jest.requireActual('../../../utils'),
  endpointFetch: jest.fn(),
  endpointPush: jest.fn(),
}));

useUserContext.mockImplementation(() => ({
  user: {},
}));

endpointFetch.mockResolvedValue({
  price: 100000,
});
endpointPush.mockResolvedValue({
  ...NEW_FINANCE_QUOTE,
  ...(NEW_FINANCE_QUOTE.HP.financeType = 'HP'),
});
endpointFetch.mockResolvedValue(DELIVERY_INFO);

useGOContext.mockImplementation(() => ({ isGoRdy: true }));
useVehicleContext.mockImplementation(() => ({ vehicle: { leadgen_B: false } }));

beforeEach(() => {
  Object.defineProperty(window, 'google_optimize', {
    value: {
      get: () => '1',
    },
    configurable: true,
    writable: true,
  });
});

useGOContext.mockImplementation(() => ({ isGoRdy: true }));

beforeEach(() => {
  Object.defineProperty(window, 'google_optimize', {
    value: {
      get: () => '1',
    },
    configurable: true,
    writable: true,
  });
});

const submitEvent = {
  preventDefault: jest.fn(),
};

describe('<Terms />', () => {
  let wrapper;
  const getWrapper = (method = shallow) => method(<Terms />);

  it('rendered without errors', async () => {
    wrapper = getWrapper();
    await global.updateWrapper(wrapper);
    expect(wrapper).toMatchSnapshot();
  });

  it('handles fields change events TermsInputs', () => {
    wrapper = getWrapper();
    const getChild = () => wrapper.find('TermsInputs');

    getChild().invoke('handleChange')({
      target: { name: 'deposit', value: '' },
    });
    expect(getChild().prop('fields').deposit.status).toEqual(
      INPUT_STATUS.INVALID
    );

    getChild().invoke('handleChange')({
      target: { name: 'deposit', value: 111100 },
    });
    getChild().invoke('handleChange')({
      target: { name: 'finance', value: 'Hire Purchase' },
    });
    expect(getChild().prop('fields').deposit.status).toEqual(
      INPUT_STATUS.INVALID
    );
    expect(getChild().prop('fields').finance.value).toEqual('HP');

    getChild().invoke('handleChange')({
      target: { name: 'deposit', value: '1111' },
    });
    expect(getChild().prop('fields').deposit.value).toEqual('1111');

    getChild().invoke('handleChange')({
      target: { name: 'mileage', value: '20000' },
    });
    expect(getChild().prop('fields').mileage.value).toEqual('20000');

    getChild().invoke('handleChange')({
      target: { name: 'terms', value: '24' },
    });
    expect(getChild().prop('fields').terms.value).toEqual('24');

    getChild().invoke('handleChange')({
      target: { name: 'delivery', value: '19' },
    });
    expect(getChild().prop('fields').delivery.value).toEqual('19');

    getChild().invoke('handleChange')({
      target: { name: 'credit', value: 'good' },
    });
    expect(getChild().prop('fields').credit.value).toEqual('good');

    getChild().invoke('handleChange')({ target: { id: 'PCP' } });
    expect(getChild().prop('fields').finance.value).toEqual('PCP');

    getChild().invoke('handleChange')({
      target: { name: 'deposit', value: 111100 },
    });
    getChild().invoke('handleChange')({ target: { id: 'Hire Purchase' } });
    expect(getChild().prop('fields').deposit.status).toEqual(
      INPUT_STATUS.INVALID
    );
    expect(getChild().prop('fields').finance.value).toEqual('HP');
  });

  it('handles fields change events TermsResults', () => {
    wrapper = getWrapper();
    const getChild = () => wrapper.find('TermsResults');

    getChild().invoke('handleChange')({
      target: { name: 'terms', value: '60' },
    });
    expect(getChild().prop('fields').terms.value).toEqual('60');

    getChild().invoke('disabledForm')({ type: 'PCP' });
    getChild().invoke('showModal')({ type: 'PCP' });
    getChild().invoke('handleChange')({
      target: { name: 'finance', value: 'PCP' },
    });
    expect(getChild().prop('fields').finance.value).toEqual('PCP');
  });

  it('endpointFetch was called on mounting', async () => {
    wrapper = getWrapper(mount);
    const prodAdvertId = { productAdvertId: '3362300' };

    await global.updateWrapper(wrapper);
    expect(endpointFetch).toBeCalledWith('userProfileInfo');
    expect(endpointFetch).toBeCalledWith('productInformation', prodAdvertId);
    expect(endpointFetch).toBeCalledWith('getDeliveryOptions', prodAdvertId);
  });

  it('invalid form can`t be submitted', async () => {
    wrapper = getWrapper();
    await global.updateWrapper(wrapper);

    wrapper.find('form').simulate('submit', submitEvent);
    expect(endpointPush).not.toHaveBeenCalledWith('-=# SOME ENDPOINT #=-');
  });

  it('valid form can be submitted', () => {
    wrapper = getWrapper();

    wrapper.find('form').simulate('submit', submitEvent);
    expect(endpointPush).toHaveBeenCalled();
  });
});
