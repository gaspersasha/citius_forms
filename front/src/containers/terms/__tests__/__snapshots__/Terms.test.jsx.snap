// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Terms /> rendered without errors 1`] = `
<div
  className="container"
>
  <Seo
    meta={
      Object {
        "description": "Buyacar online with no hussle",
        "page_title": "BuyaCar: the easy way to get a great car deal",
        "rights": "Autovia Group Limited, licensed by Felden",
        "twitter": Object {
          "card": "photo",
          "site": "@buy_a_car",
          "title": "BuyaCar: the easy way to get a great car deal",
          "url": "https://website",
        },
      }
    }
  />
  <div
    className="controls"
  >
    <div
      className="align-corner"
    >
      <button
        className="back"
        type="button"
      >
        Back
      </button>
      <span
        className="progress"
      >
        0% Completed
      </span>
    </div>
    <ProgressBar
      progress={0}
    />
    <Parts
      part={2}
      step={4}
    />
  </div>
  <div
    className="form-container"
  >
    <div
      className="form"
    >
      <h1
        className="title"
      >
        Choose your finance terms
      </h1>
      <p
        className="subTitle"
      >
        Tailor your finance details to see how much you could be paying
      </p>
      <form
        className="financeForm"
        onSubmit={[Function]}
      >
        <TermsInputs
          deliveryOptions={Array []}
          disabledTrue={true}
          fields={
            Object {
              "credit": Object {
                "status": "default",
                "validationType": "notEmpty",
                "value": "",
              },
              "delivery": Object {
                "status": "default",
                "validationType": "number",
                "value": 18,
              },
              "deposit": Object {
                "status": "default",
                "validationType": "number",
                "value": "",
              },
              "finance": Object {
                "status": "default",
                "validationType": "notEmpty",
                "value": "PCP",
              },
              "mileage": Object {
                "status": "default",
                "validationType": "number",
                "value": "",
              },
              "terms": Object {
                "status": "default",
                "validationType": "number",
                "value": "",
              },
            }
          }
          handleChange={[Function]}
          isLeadGen={false}
          maxFinanceDeposit={4400}
        />
        <div
          className="financeType"
        >
          <h5>
            Now select your finance type
          </h5>
          <TermsResults
            apiError=""
            disabledForm={[Function]}
            fields={
              Object {
                "credit": Object {
                  "status": "default",
                  "validationType": "notEmpty",
                  "value": "",
                },
                "delivery": Object {
                  "status": "default",
                  "validationType": "number",
                  "value": 18,
                },
                "deposit": Object {
                  "status": "default",
                  "validationType": "number",
                  "value": "",
                },
                "finance": Object {
                  "status": "default",
                  "validationType": "notEmpty",
                  "value": "PCP",
                },
                "mileage": Object {
                  "status": "default",
                  "validationType": "number",
                  "value": "",
                },
                "terms": Object {
                  "status": "default",
                  "validationType": "number",
                  "value": "",
                },
              }
            }
            handleChange={[Function]}
            quotes={Object {}}
            showModal={[Function]}
          />
        </div>
        <Button
          className="submitBtn"
          disabled={true}
          href=""
          onClick={null}
          selected={false}
          styleType="primary"
          type="submit"
        >
          Next
        </Button>
      </form>
      <div
        className="descriptions"
      >
        <p>
          Representative example when buying on PCP
        </p>
        <p>
          Borrowing £9,500 over 48 months, zero deposit, on PCP, an annual mileage of 8,000pa, with a Representative APR of 9.9%, the amount payable would be £192 per month, an optional final payment of £2,923, with a total cost of credit of £2,426 and a total amount payable of £11,926 (** based on a Ford Fiesta Titanium hatchback)
        </p>
        <p>
          Buyacar is a credit broker, not a lender. Our rates start from 6.9% APR. The rate you are offered will depend on your individual circumstances.
        </p>
      </div>
    </div>
    <PrivacyNotice
      onChange={[Function]}
      unsubscribe={false}
    />
  </div>
</div>
`;
